pkg/gather/gather.go

```diff
@@ -0,0 +1,244 @@
+			gathererName, result.FunctionName, result.TimeElapsed, len(result.Records),
+		)
+
+		functionReports = append(functionReports, GathererFunctionReport{
+			FuncName:     fmt.Sprintf("%v/%v", gathererName, result.FunctionName),
+			Duration:     result.TimeElapsed.Milliseconds(),
+			RecordsCount: len(result.Records),
+			Errors:       errorsToStrings(result.Errs),
+			Panic:        result.Panic,
+		})
+	}
+
+	return functionReports, sumErrors(errs)
+}
+
+// RecordArchiveMetadata records info about archive and gatherers' reports
+func RecordArchiveMetadata(
+	functionReports []GathererFunctionReport,
+	rec recorder.Interface,
+	anonymizer *anonymization.Anonymizer,
+) error {
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+
+	archiveMetadata := record.Record{
+		Name: archiveMetadataFilename,
+		Item: record.JSONMarshaller{Object: ArchiveMetadata{
+			StatusReports:              functionReports,
+			MemoryAlloc:                m.HeapAlloc,
+			Uptime:                     time.Since(programStartTime).Truncate(time.Millisecond).Seconds(),
+			IsGlobalObfuscationEnabled: anonymizer != nil,
+		}},
+	}
+	if err := rec.Record(archiveMetadata); err != nil {
+		return fmt.Errorf("unable to record archive metadata because of error: %v", err)
+	}
+
+	return nil
+}
+
+// startGatheringConcurrently starts gathering of enabled functions of the provided gatherer and returns a channel
+// with results which will be closed when processing is done
+func startGatheringConcurrently(
+	ctx context.Context, gatherer gatherers.Interface, enabledFunctions []string,
+) (chan GatheringFunctionResult, error) {
+	gathererName := gatherer.GetName()
+
+	gatherAllFunctions, gatherFunctionsList := getListOfEnabledFunctionForGatherer(
+		gathererName, enabledFunctions,
+	)
+	if !gatherAllFunctions && len(gatherFunctionsList) == 0 {
+		return nil, fmt.Errorf("no gather functions are specified to run")
+	}
+
+	var tasks []Task
+
+	for functionName, gatheringClosure := range gatherer.GetGatheringFunctions() {
+		if !gatherAllFunctions && !utils.StringInSlice(functionName, gatherFunctionsList) {
+			continue
+		}
+
+		tasks = append(tasks, Task{
+			Name: functionName,
+			F:    gatheringClosure,
+		})
+	}
+
+	return HandleTasksConcurrently(ctx, tasks), nil
+}
+
+// getListOfEnabledFunctionForGatherer parses a list of gathering functions to enable,
+// which has the following structure: []string {
+//   "clusterconfig/container_images",
+//   "clusterconfig/nodes",
+//   "clusterconfig/authentication",
+//   "othergatherer/some_function",
+// } where each item consists of a gatherer name and a function name split by a slash.
+// If there's a string "ALL", we enable everything and return the first parameter as true,
+// otherwise it will be false and the second parameter will contain function names
+func getListOfEnabledFunctionForGatherer(gathererName string, allFunctionsList []string) (bool, []string) {
+	if utils.StringInSlice(AllGatherersConst, allFunctionsList) {
```

Comment:
Is there any way to disable the gatherer as well? Right now we only support disabling the gathers.

Also... Wouldn't it be better if we assume that we don't specify the gathers per configuration we always run all the gathers? Let me try to explain it.

Right now, we should have at the configuration file the value `ALL` like this:

```yml
gather:
  - ALL
```

And If the configuration is missing we run all the gathers? Also, don't we want to specify the gathers per gatherer when disabling it?

```yml
gather:
    clusterconfig:
      - image_registries
      - image_pruners
      - install_plans
```
In this case we will run only the `clusterconfig` with the listed gathers.

Reply:
Right now we specify only enabled gathering functions, we can enable the whole gatherer, but we'd need to mention every function. I thought of making it lil more flexible, but decided to not make the PR even bigger. I'd vote for possibly improving it later.

Reply:
Sure! We should have another PR to implement it. I'm just rising the discussion here, and maybe wasn't the best place, but it was easy to explain by adding the comment here.